core/src/main/java/io/nixer/nixerplugin/core/detection/rules/ratio/FailedLoginRatioRule.java

```diff
@@ -15,58 +13,58 @@
         Assert.notNull(loginMetric, "LoginMetric must not be null");
         this.loginMetric = loginMetric;
 
-        Assert.notNull(activationLevel, "activationLevel must not be null");
-        Assert.notNull(activationLevel, "deactivationLevel must not be null");
         if (activationLevel < deactivationLevel) {
             throw new IllegalStateException(String.format("Activation level (%d) must be equal or bigger than deactivation level (%d)",
                     activationLevel,
                     deactivationLevel));
         }
-        this.activationLevel = new AtomicInteger(activationLevel);
-        this.deactivationLevel = new AtomicInteger(deactivationLevel);
-
-        Assert.notNull(minimumSampleSize, "minimumSampleSize must not be null");
-        this.minimumSampleSize = new AtomicInteger(minimumSampleSize);
+        if (activationLevel < 0 || activationLevel > 100 || deactivationLevel < 0 || deactivationLevel > 100) {
+            throw new IllegalStateException(String.format("Activation (%d) and deactivation (%d) levels must be within 0-100 range",
+                    activationLevel,
+                    deactivationLevel));
+        }
+        this.activationLevel = activationLevel;
+        this.deactivationLevel = deactivationLevel;
+        this.minimumSampleSize = minimumSampleSize;
     }
 
     @Override
     public void execute(final LoginContext context, final EventEmitter eventEmitter) {
         final int successCount = loginMetric.value(LoginResult.Status.SUCCESS.getName());
         final int failureCount = loginMetric.value(LoginResult.Status.FAILURE.getName());
-        final int minimumSample = minimumSampleSize.get();
-        final int activation = activationLevel.get();
-        final int deactivation = deactivationLevel.get();
 
-        if (successCount + failureCount < minimumSample || (successCount == 0 && failureCount == 0)) {
+        if (successCount + failureCount < minimumSampleSize || (successCount == 0 && failureCount == 0)) {
             return;
         }
         if (successCount > 0 && failureCount == 0) {
             eventEmitter.accept(new FailedLoginRatioDeactivationEvent(0));
             return;
         }
         if (failureCount > 0 && successCount == 0) {
-            eventEmitter.accept(new FailedLoginRatioActivationEvent(100));
+            eventEmitter.accept(new FailedLoginRatioActivationEvent(1));
             return;
         }
 
-        double ratio = ((double) failureCount / (failureCount + successCount)) * 100;
-        logger.debug("Calculated failed login ratio: " + ratio);
+        double ratio = ((double) failureCount / (failureCount + successCount));
+        if (logger.isDebugEnabled()) {
+            logger.debug("Calculated failed login ratio: " + ratio);
+        }
 
-        if (ratio >= activation) {
+        if (ratio * 100 >= activationLevel) {
```

Comment:
It would be nice to extract the `100` multiplier to some constant named in descriptive way. Now it's just a magic number.

Reply:
Not sure about that. I added comments.

Reply:
Ok, as discussed offline.